src/Place/ReadModel/History/HistoryProjector.php

```diff
@@ -0,0 +1,218 @@
+<?php declare(strict_types=1);
+
+namespace CultuurNet\UDB3\Place\ReadModel\History;
+
+use Broadway\Domain\DateTime as BroadwayDateTime;
+use Broadway\Domain\DomainMessage;
+use Broadway\Domain\Metadata;
+use Broadway\EventHandling\EventListenerInterface;
+use CultuurNet\UDB3\Cdb\EventItemFactory;
+use CultuurNet\UDB3\Event\ReadModel\DocumentRepositoryInterface;
+use CultuurNet\UDB3\Event\ReadModel\History\Log;
+use CultuurNet\UDB3\Place\Events\DescriptionTranslated;
+use CultuurNet\UDB3\Place\Events\LabelAdded;
+use CultuurNet\UDB3\Place\Events\LabelRemoved;
+use CultuurNet\UDB3\Place\Events\PlaceCreated;
+use CultuurNet\UDB3\Place\Events\PlaceDeleted;
+use CultuurNet\UDB3\Place\Events\PlaceImportedFromUDB2;
+use CultuurNet\UDB3\Place\Events\PlaceUpdatedFromUDB2;
+use CultuurNet\UDB3\Place\Events\TitleTranslated;
+use CultuurNet\UDB3\Place\ReadModel\Enum\EventDescription;
+use CultuurNet\UDB3\ReadModel\JsonDocument;
+use DateTime;
+use DateTimeZone;
+
+final class HistoryProjector implements EventListenerInterface
+{
+    /**
+     * @var DocumentRepositoryInterface
+     */
+    private $documentRepository;
+
+    public function __construct(DocumentRepositoryInterface $documentRepository)
+    {
+        $this->documentRepository = $documentRepository;
+    }
+
+    public function handle(DomainMessage $domainMessage)
```

Comment:
The way the `handle` method is implemented is very different from https://github.com/cultuurnet/udb3-silex/blob/master/src/Event/ReadModel/History/HistoryProjector.php#L42 which may be a bit confusing in the long run.

Additionally most events could be handled by the `createOfferEventHandler` in that history projector, so maybe we could extract some of that shared logic to a base class / trait / separate class we can inject?

It would also make it easier to add more data from the metadata in the future, as we will now have to do that twice with the `get...FromMetadata` methods in both classes.

Reply:
I have extracted a `BaseHIstoryProjector` and put some shared functionality there. 

As for the `createOfferEventHandler` method. We would have to modify it to be used with Place events since they have `getPlaceId()` method instead of `getEventId()`.

The approach I used in "my" handler does require more typing since there is one method for each event and that method is just a proxy method to the `writeHistory` method. But, here I guess we're entering the highly opinionated waters, to me, it feels more explicit when coming to code to see what is happening.

So, for now, I haven't extracted the `createOfferEventHandler` but we can discuss it some more ofc :))

Reply:
> We would have to modify it to be used with Place events since they have getPlaceId() method instead of getEventId().

`createOfferEventHandler` uses `getItemId()` which is also what all of your handlers use except for `projectPlaceCreated` 😉 

Most of the events handled in both history projectors will be inheriting from `CultuurNet\UDB3\Offer\Events\AbstractEvent`, which uses `getItemId()`.

Only events specific to places or events will be extending from other classes which use `getPlaceId()` or `getEventId()`. But that's a minority.